lib/metadata/client-state.js

```diff
@@ -0,0 +1,96 @@
+'use strict';
+
+var util = require('util');
+var errors = require('../errors');
+
+/**
+ * Creates a new instance of <code>ClientState</code>.
+ * @classdesc
+ * Represents the state of a {@link Client}.
+ * <p>
+ * Exposes information on the connections maintained by a Client at a specific time.
+ * </p>
+ * @alias module:metadata~ClientState
+ * @param {Array<Host>} hosts
+ * @param {Object.<String, Number>} openConnections
+ * @param {Object.<String, Number>} inFlightQueries
+ * @constructor
+ */
+function ClientState(hosts, openConnections, inFlightQueries) {
+  this._hosts = hosts;
+  this._openConnections = openConnections;
+  this._inFlightQueries = inFlightQueries;
+}
+
+/**
+ * Creates a new instance from the provided client.
+ * @param {Client} client
+ * @internal
+ * @ignore
+ */
+ClientState.from = function (client) {
+  var openConnections = {};
+  var inFlightQueries = {};
+  var hostArray = [];
+  client.hosts.forEach(function each(host) {
+    if (host.pool.connections.length === 0) {
+      return;
+    }
+    hostArray.push(host);
+    openConnections[host.address] = host.pool.connections.length;
+    inFlightQueries[host.address] = host.pool.connections.reduce(function (count, connection) {
+      return count + connection.getInFlight();
+    }, 0);
+  });
+  return new ClientState(hostArray, openConnections, inFlightQueries);
+};
+
+/**
+ * Get an array of hosts to which the client is connected to.
+ * @return {Array<Host>}
+ */
+ClientState.prototype.getConnectedHosts = function () {
+  return this._hosts;
+};
+
+/**
+ * Gets the amount of open connections to a given host.
+ * @param {Host} host
+ * @return {Number|undefined}
```

Comment:
One question here.  Say you have a previous reference to a host.  The host goes down, the user requests new state and then calls this method.  In this case `undefined` is returned.  Wondering if it would be better to return `0` in this case (and also for `getInFlightQueries`) or if you think `undefined` is better since it communicates to user that the host is not connected, thus there is no state for it.

Reply:
I was using `undefined` as in "We don't have that info", but now that I think about it, for reporting (like logging / tracing info) it would be simpler for the user if we always return a number. I will change it.

Reply:
👍 that was my thinking as well.  I had some code that was iterating over all hosts (not just connected ones) and encountered that, now I know longer need to special case if returns undefined so the change should work great.